visualisation/grouped_spectra.py

```diff
@@ -0,0 +1,92 @@
+    
+    fig_grouped_corr = go.Figure()
+    
+    if spectra_conversion_type == RAW:
+        file_name += '_raw'
+        shift = st.slider(SHIFT, 0, 30000, 0, 250)
+        col1, col2 = st.beta_columns((2))
+        for col in range(len(df.columns)):
+            col_name = df.columns[col]
+            
+            corrected = pd.DataFrame(df.loc[:, col_name]).dropna()
+            
+            df_to_save[col_name] = corrected[col_name]
+            
+            if col != 0:
+                corrected.iloc[:, 0] += shift * col
+            
+            fig_grouped_corr = draw.add_traces(corrected.reset_index(), fig_grouped_corr, x=RS, y=col_name,
+                                               name=f'{df.columns[col]}')
+        draw.fig_layout(template, fig_grouped_corr, plots_colorscale=plots_color, descr=ORG)
+    
+    
+    
+    elif spectra_conversion_type == OPT or spectra_conversion_type == NORM:
+        file_name += '_optimized'
+        df2 = df.copy()
+        df_to_save = pd.DataFrame()
+
+        if spectra_conversion_type == OPT:
+            shift = st.slider(SHIFT, 0, 30000, 0, 250)
+
+        elif spectra_conversion_type == NORM:
+            file_name += '_normalized'
+            shift = st.slider(SHIFT, 0.0, 1.0, 0.0, 0.1)
+
+        adjust_plots_globally = st.radio(
+            "Adjust all spectra or each spectrum?",
+            ('all', 'each'), index=0)
+        col1, col2 = st.beta_columns((2, 1))
+        with col2:
+            st.markdown('## Adjust your spectra')
+    
+            if adjust_plots_globally == 'all':
+                deg, window = utils.adjust_spectras_by_window_and_degree()
+                vals = {col: (deg, window) for col in df.columns}
+    
+            elif adjust_plots_globally == 'each':
+                with st.beta_expander("Customize your chart"):
+                    vals = {col: utils.adjust_spectras_by_window_and_degree(col) for col in df.columns}
+
+        for col_ind, col in enumerate(df2.columns):
+            corrected = utils.process_grouped_opt_spec(df2=df2,
+                                                       spectra_conversion_type=spectra_conversion_type,
+                                                       col=col,
+                                                       deg=vals[col][0],
+                                                       window=vals[col][1])
+            df_to_save[col] = corrected[col]
+    
+            if col_ind != 0:
+                corrected.iloc[:, 0] += shift * col_ind
+    
+            fig_grouped_corr = draw.add_traces(corrected.reset_index(), fig_grouped_corr, x=RS, y=col,
+                                               name=col)
+            draw.fig_layout(template, fig_grouped_corr, plots_colorscale=plots_color, descr=OPT_S)
+    with col1:
+        st.write(fig_grouped_corr)
+
+    save_read.save_adj_spectra_to_file(df_to_save, file_name)
```

Comment:
Cała ta funkcja jest za duża, za skomplikowana i robi za dużo.
Ja jeszcze nad tym pomyślę i jak chcesz to ja ją poprawię, ale widzę to mniej wiecej tak.

1. jedna funkcja powinna zwracać to co nazywasz df_to_save
2. potem mając df_to_save można je przekazać do funkcji plotującej, która zrobi z tego wykres (brałaby shift jako argument)
3. potem mając df_to_save można by je przekazać do `save_read.save_adj_spectra_to_file` tylko raczej w `vis.py` bo to pokazuje co w zasadzie będzie się działo 

I jest tam jakiś bałagan z tym `col1, col2`

W ogólności powinniśmy separować funkcje które coś robią (operacje na danych, rysowanie wykresów) od tego, że pokazujemy coś w streamlit. 
Wywołania typu st.cośtam powinny być tylko w vis.py albo jakimś innym pomocniczym module - tak żebyśmy mogli też korzystać z przetwarzania danych bez używania streamlita.

Reply:
To jest super plan. ;] szczegolnie ostatni i pierwszy